Const enum initializer type

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.3.0-dev.201xxxxx

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**
`const enum initializer` `ts2474`
**Code**

```ts
// lib.ts
const enum STATIC_KEYS {
    ZERO,
    ONE,
    TWO,

    LAST
}

export const STARTING_NUMBER = STATIC_KEYS.LAST;

// consumer.ts
const enum MORE_STATIC_KEYS {
    ONE = STARTING_NUMBER, // error TS2474; STATIC_KEYS.LAST works as expected
    TWO,
    THREE,
}
```

**Expected behavior:**
Code compiles with `MORE_STATIC_KEYS.ONE === STATIC_KEYS.LAST`.

**Actual behavior:**
`error TS2474: In 'const' enum declarations member initializer must be constant expression.`
TS recognizes that `STARTING_NUMBER` has the type `STATIC_KEYS.LAST`, but doesn't consider it a 'constant expression.'

If this is expected behavior, clarity in the error message would be appreciated.

**Playground Link:** [link](http://www.typescriptlang.org/play/#src=%2F%2F%20lib.ts%0D%0Aconst%20enum%20STATIC_KEYS%20%7B%0D%0A%20%20%20%20ZERO%2C%0D%0A%20%20%20%20ONE%2C%0D%0A%20%20%20%20TWO%2C%0D%0A%0D%0A%20%20%20%20LAST%0D%0A%7D%0D%0A%0D%0Aexport%20const%20STARTING_NUMBER%20%3D%20STATIC_KEYS.LAST%3B%0D%0A%0D%0A%2F%2F%20consumer.ts%0D%0Aconst%20enum%20MORE_STATIC_KEYS%20%7B%0D%0A%20%20%20%20ONE%20%3D%20STARTING_NUMBER%2C%20%2F%2F%20STATIC_KEYS.LAST%20works%20as%20expected%0D%0A%20%20%20%20TWO%2C%0D%0A%20%20%20%20THREE%2C%0D%0A%7D%0D%0A%0D%0A)

**Related Issues:** <!-- Did you find other bugs that looked similar? -->
Multiple discussions about merging/intersecting of const enums, of which the above is my attempted workaround.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -26903,7 +26903,7 @@ namespace ts {
                 return 0;
             }
             else if (isConstEnum) {
-                error(initializer, Diagnostics.In_const_enum_declarations_member_initializer_must_be_constant_expression);
+                error(initializer, Diagnostics.const_enum_member_initializers_can_only_contain_literal_values_and_other_const_enums_values);
             }
             else if (member.parent.flags & NodeFlags.Ambient) {
                 error(initializer, Diagnostics.In_ambient_enum_declarations_member_initializer_must_be_constant_expression);

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -26903,7 +26903,7 @@ namespace ts {
                 return 0;
             }
             else if (isConstEnum) {
-                error(initializer, Diagnostics.const_enum_member_initializers_can_only_contain_literal_values_and_other_const_enums_values);
+                error(initializer, Diagnostics.const_enum_member_initializers_can_only_contain_literal_values_and_other_computed_enum_values);
             }
             else if (member.parent.flags & NodeFlags.Ambient) {
                 error(initializer, Diagnostics.In_ambient_enum_declarations_member_initializer_must_be_constant_expression);

--- a/src/compiler/diagnosticMessages.json
+++ b/src/compiler/diagnosticMessages.json
@@ -1696,7 +1696,7 @@
         "category": "Error",
         "code": 2473
     },
-    "In 'const' enum declarations member initializer must be constant expression.": {
+    "const enum member initializers can only contain literal values and other const enums values.": {
         "category": "Error",
         "code": 2474
     },

--- a/src/compiler/diagnosticMessages.json
+++ b/src/compiler/diagnosticMessages.json
@@ -1696,7 +1696,7 @@
         "category": "Error",
         "code": 2473
     },
-    "const enum member initializers can only contain literal values and other const enums values.": {
+    "const enum member initializers can only contain literal values and other computed enum values.": {
         "category": "Error",
         "code": 2474
     },

--- a/tests/baselines/reference/constEnum2.errors.txt
+++ b/tests/baselines/reference/constEnum2.errors.txt
@@ -1,7 +1,7 @@
-tests/cases/conformance/constEnums/constEnum2.ts(10,9): error TS2474: In 'const' enum declarations member initializer must be constant expression.
-tests/cases/conformance/constEnums/constEnum2.ts(11,9): error TS2474: In 'const' enum declarations member initializer must be constant expression.
+tests/cases/conformance/constEnums/constEnum2.ts(10,9): error TS2474: const enum member initializers can only contain literal values and other const enums values.
+tests/cases/conformance/constEnums/constEnum2.ts(11,9): error TS2474: const enum member initializers can only contain literal values and other const enums values.
 tests/cases/conformance/constEnums/constEnum2.ts(12,5): error TS1005: ',' expected.
-tests/cases/conformance/constEnums/constEnum2.ts(12,9): error TS2474: In 'const' enum declarations member initializer must be constant expression.
+tests/cases/conformance/constEnums/constEnum2.ts(12,9): error TS2474: const enum member initializers can only contain literal values and other const enums values.
 
 
 ==== tests/cases/conformance/constEnums/constEnum2.ts (4 errors) ====
@@ -16,13 +16,13 @@ tests/cases/conformance/constEnums/constEnum2.ts(12,9): error TS2474: In 'const'
         d = 10,
         e = 199 * Math.floor(Math.random() * 1000),
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2474: In 'const' enum declarations member initializer must be constant expression.
+!!! error TS2474: const enum member initializers can only contain literal values and other const enums values.
         f = d - (100 * Math.floor(Math.random() % 8))
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2474: In 'const' enum declarations member initializer must be constant expression.
+!!! error TS2474: const enum member initializers can only contain literal values and other const enums values.
         g = CONST,
         ~
 !!! error TS1005: ',' expected.
             ~~~~~
-!!! error TS2474: In 'const' enum declarations member initializer must be constant expression.
+!!! error TS2474: const enum member initializers can only contain literal values and other const enums values.
     }

--- a/tests/baselines/reference/constEnum2.errors.txt
+++ b/tests/baselines/reference/constEnum2.errors.txt
@@ -1,7 +1,7 @@
-tests/cases/conformance/constEnums/constEnum2.ts(10,9): error TS2474: const enum member initializers can only contain literal values and other const enums values.
-tests/cases/conformance/constEnums/constEnum2.ts(11,9): error TS2474: const enum member initializers can only contain literal values and other const enums values.
+tests/cases/conformance/constEnums/constEnum2.ts(10,9): error TS2474: const enum member initializers can only contain literal values and other computed enum values.
+tests/cases/conformance/constEnums/constEnum2.ts(11,9): error TS2474: const enum member initializers can only contain literal values and other computed enum values.
 tests/cases/conformance/constEnums/constEnum2.ts(12,5): error TS1005: ',' expected.
-tests/cases/conformance/constEnums/constEnum2.ts(12,9): error TS2474: const enum member initializers can only contain literal values and other const enums values.
+tests/cases/conformance/constEnums/constEnum2.ts(12,9): error TS2474: const enum member initializers can only contain literal values and other computed enum values.
 
 
 ==== tests/cases/conformance/constEnums/constEnum2.ts (4 errors) ====
@@ -16,13 +16,13 @@ tests/cases/conformance/constEnums/constEnum2.ts(12,9): error TS2474: const enum
         d = 10,
         e = 199 * Math.floor(Math.random() * 1000),
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2474: const enum member initializers can only contain literal values and other const enums values.
+!!! error TS2474: const enum member initializers can only contain literal values and other computed enum values.
         f = d - (100 * Math.floor(Math.random() % 8))
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2474: const enum member initializers can only contain literal values and other const enums values.
+!!! error TS2474: const enum member initializers can only contain literal values and other computed enum values.
         g = CONST,
         ~
 !!! error TS1005: ',' expected.
             ~~~~~
-!!! error TS2474: const enum member initializers can only contain literal values and other const enums values.
+!!! error TS2474: const enum member initializers can only contain literal values and other computed enum values.
     }

--- a/tests/baselines/reference/constEnumErrors.errors.txt
+++ b/tests/baselines/reference/constEnumErrors.errors.txt
@@ -1,8 +1,8 @@
 tests/cases/compiler/constEnumErrors.ts(1,12): error TS2567: Enum declarations can only merge with namespace or other enum declarations.
 tests/cases/compiler/constEnumErrors.ts(5,8): error TS2567: Enum declarations can only merge with namespace or other enum declarations.
 tests/cases/compiler/constEnumErrors.ts(12,9): error TS2651: A member initializer in a enum declaration cannot reference members declared after it, including members defined in other enums.
-tests/cases/compiler/constEnumErrors.ts(14,9): error TS2474: const enum member initializers can only contain literal values and other const enums values.
-tests/cases/compiler/constEnumErrors.ts(15,10): error TS2474: const enum member initializers can only contain literal values and other const enums values.
+tests/cases/compiler/constEnumErrors.ts(14,9): error TS2474: const enum member initializers can only contain literal values and other computed enum values.
+tests/cases/compiler/constEnumErrors.ts(15,10): error TS2474: const enum member initializers can only contain literal values and other computed enum values.
 tests/cases/compiler/constEnumErrors.ts(22,13): error TS2476: A const enum member can only be accessed using a string literal.
 tests/cases/compiler/constEnumErrors.ts(24,13): error TS2476: A const enum member can only be accessed using a string literal.
 tests/cases/compiler/constEnumErrors.ts(26,9): error TS2475: 'const' enums can only be used in property or index access expressions or the right hand side of an import declaration or export assignment or type query.
@@ -35,10 +35,10 @@ tests/cases/compiler/constEnumErrors.ts(42,9): error TS2478: 'const' enum member
         // forward reference to the element of the same enum
         Y = E1.Z,
             ~~~~
-!!! error TS2474: const enum member initializers can only contain literal values and other const enums values.
+!!! error TS2474: const enum member initializers can only contain literal values and other computed enum values.
         Y1 = E1["Z"]
              ~~~~~~~
-!!! error TS2474: const enum member initializers can only contain literal values and other const enums values.
+!!! error TS2474: const enum member initializers can only contain literal values and other computed enum values.
     }
     
     const enum E2 {

--- a/tests/baselines/reference/constEnumErrors.errors.txt
+++ b/tests/baselines/reference/constEnumErrors.errors.txt
@@ -1,8 +1,8 @@
 tests/cases/compiler/constEnumErrors.ts(1,12): error TS2567: Enum declarations can only merge with namespace or other enum declarations.
 tests/cases/compiler/constEnumErrors.ts(5,8): error TS2567: Enum declarations can only merge with namespace or other enum declarations.
 tests/cases/compiler/constEnumErrors.ts(12,9): error TS2651: A member initializer in a enum declaration cannot reference members declared after it, including members defined in other enums.
-tests/cases/compiler/constEnumErrors.ts(14,9): error TS2474: In 'const' enum declarations member initializer must be constant expression.
-tests/cases/compiler/constEnumErrors.ts(15,10): error TS2474: In 'const' enum declarations member initializer must be constant expression.
+tests/cases/compiler/constEnumErrors.ts(14,9): error TS2474: const enum member initializers can only contain literal values and other const enums values.
+tests/cases/compiler/constEnumErrors.ts(15,10): error TS2474: const enum member initializers can only contain literal values and other const enums values.
 tests/cases/compiler/constEnumErrors.ts(22,13): error TS2476: A const enum member can only be accessed using a string literal.
 tests/cases/compiler/constEnumErrors.ts(24,13): error TS2476: A const enum member can only be accessed using a string literal.
 tests/cases/compiler/constEnumErrors.ts(26,9): error TS2475: 'const' enums can only be used in property or index access expressions or the right hand side of an import declaration or export assignment or type query.
@@ -35,10 +35,10 @@ tests/cases/compiler/constEnumErrors.ts(42,9): error TS2478: 'const' enum member
         // forward reference to the element of the same enum
         Y = E1.Z,
             ~~~~
-!!! error TS2474: In 'const' enum declarations member initializer must be constant expression.
+!!! error TS2474: const enum member initializers can only contain literal values and other const enums values.
         Y1 = E1["Z"]
              ~~~~~~~
-!!! error TS2474: In 'const' enum declarations member initializer must be constant expression.
+!!! error TS2474: const enum member initializers can only contain literal values and other const enums values.
     }
     
     const enum E2 {

